src/data.js

```diff
@@ -263,45 +263,154 @@ let Artifact = Entity.configure({
      *
      * storageType: error
      *   reason:        Formalized reason for error artifact, see JSON schema.
      *   message:       Human readable error message to return
+     *
+     * storageType: s3 (deprecated)
+     *   bucket:        S3 bucket that contains the object
+     *   prefix:        Prefix (path) for the object within the bucket
+     *
+     * storageType: azure (deprecated)
+     *   container:     Azure container that holds the blob
+     *   path:          Path to blob within container
+     *
      */
-    details:        Entity.types.JSON,
-    expires:        Entity.types.Date,
+    details: Entity.types.JSON,
+    expires: Entity.types.Date,
+    present: Entity.types.Number, // 0 or 1 to represent whether the artifact is there
+  },
+  migrate: function (item) {
+    // Add the present flag.  We will set this value to 1 for all cases because
+    // they are either stored in this table and can be worked on atomically so
+    // are always present if the entity can be loaded or because they are the
+    // deprecated s3 or azure storageType which will never have checking for
+    // presence
+    return {
+      taskId: item.taskId,
+      runId: item.runId,
+      name: item.name,
+      storageType: item.storageType,
+      contentType: item.contentType,
+      details: item.details,
+      expires: item.expires,
+      present: 1,
+    };
   },
   context: [
-    'blobStore',      // BlobStore instance wrapping Azure Blob Storage
-    'privateBucket',  // Private artifact bucket wrapping S3
-    'publicBucket',   // Public artifact bucket wrapping S3
-    'monitor',        // base.monitor instance
+    'blobStore', // BlobStore instance wrapping Azure Blob Storage
+    'privateBucket', // Private artifact bucket wrapping S3
+    'publicBucket', // Public artifact bucket wrapping S3
+    'monitor', // base.monitor instance
+    's3', // Mapping of aws region to aws sdk object
   ],
 });
 
 /** Return JSON representation of artifact meta-data */
 Artifact.prototype.json = function() {
   return {
-    storageType:      this.storageType,
-    name:             this.name,
-    expires:          this.expires.toJSON(),
-    contentType:      this.contentType,
+    storageType: this.storageType,
+    name: this.name,
+    expires: this.expires.toJSON(),
+    contentType: this.contentType,
   };
 };
 
+// Make sure that our hash digests are valid hex
+let validMD5 = /^[a-fA-F0-9]{32}$/;
+let validSHA256 = /^[a-fA-F0-9]{64}$/;
+
+/**
+ * A light wrapper around the create function that ensures some basic things
+ * are correct about the blobs, as well as forcing the present property to be 1
+ * for non-blob storageTypes and converting the truthiness of the present
+ * property to 0 or 1
+ *
+ * NOTE: I have included code here in comments which should be uncommented
+ * when we wish to bar creation of new s3 or azure entities
+ */
+Artifact.create = function(properties, overwriteIfExists /*, allowDeprecated=false*/) {
```

Comment:
I suggest that we don't do this...

If you really want validation I propose that we add a new type to azure-entities.
Instead of `Entity.types.JSON` we could do: `Entity.types.Schema({type: object, ...})`, and then let the type wrap a JSON schema to ensure validation...